### PR TITLE
Fix watchKv+backoff bug:

### PR DESCRIPTION
**Pull request by advait**
## Context:
  - The backoff.RetryNotify API assumes that the operation func returns reasonably quickly.
  - Long running funcs (e.g. infinite loops like watchKv) end up eventually exceeding the ExponentialBackoff.MaxElapsedTime threshold. When the underlying func fails, RetryNotify thinks that it has exceeded the MaxElapsedTime and simply exits without backing off.
  - See offending backoff code [here](https://github.com/cenk/backoff/blob/32cd0c5b3aef12c76ed64aaf678f6c79736be7dc/exponential.go?ts=2#L120)

## Symptoms:
  - We had traefik running smoothly for several days (watchKv operation running for a *long* time). We had a network blip that resulted in the watch kv chan closing. This caused traefik to stop watching the key when it should have backoff/retried instead.

## The Fix:
  - This changes preserves the semantics of watchKv while converting the
    backoff operation func into a quick-returning one.

## Broader impact:
  - @emilevauge @vdemeester offhand can you think of other situations where we may have long-running (infinite looping) funcs that are wrapped with backoff? They are symptomatic of these flaws as well.

### CODE REVIEW

File: provider/kv.go [hunk 1]
```diff
@@ -60,17 +86,19 @@ func (provider *Kv) watchKv(configurationChan chan<- types.ConfigMessage, prefix
...
-		log.Errorf("KV connection error: %+v, retrying in %s", err, time)
-	}
-	err := backoff.RetryNotify(operation, backoff.NewExponentialBackOff(), notify)
-	if err != nil {
-		return fmt.Errorf("Cannot connect to KV server: %v", err)
+		notify := func(err error, time time.Duration) {
+			log.Errorf("KV watch connection error: %+v, retrying in %s", err, time)
+		}
+		b := backoff.NewExponentialBackOff()
+		b.MaxElapsedTime = 0 // Keep backing off forever
```
Comment (advait) [inline provider/kv.go]: This change is a bit controversial. My feelings: In the event that an etcd cluster goes down (or there's a network partition, etc.) it's going to take us longer than `DefaultMaxElapsedTime` (10 minutes) to get the cluster back up. Given that we're logging with the `notify` func here, is it harmful to have traefik backoff/retry indefinitely?

**Comment by advait** [comment]
I'm having trouble getting `make crossbinary` to work on Travis. It's working for me just fine locally. Any tips?

**Comment by vdemeester** [comment]
> I'm having trouble getting `make crossbinary` to work on Travis. It's working for me just fine locally. Any tips?
@advait you're not the only one :wink: @emilevauge is one it I think :stuck_out_tongue_closed_eyes:

**Comment by vdemeester** [comment]
Hum I think other provider might have the same problem :sweat:

**Comment by emilevauge** [comment]
Hi @advait, thanks a lot !
There is currently an issue with travis on go cross compilations. I need to investigate on it before reviewing your PR.

**Comment by emilevauge** [comment]
You can now rebase to master to get travis working again ;) The issue was in https://github.com/mitchellh/gox.

**Comment by emilevauge** [comment]
@advait thanks for reporting this issue.
As @vdemeester said, it also affects other providers (Docker, Marathon, Consul Services...).
Don't you think it would be simpler to create a generic wrapper around the backoff library that would manage long running jobs? We could then reuse it in all providers.

**Comment by advait** [comment]
@emilevauge Sorry for the late response. I don't think that is possible. A retry-able operation has two final states: success or failure. An infinite loop is this weird quasi-state that is neither success nor failure. The backoff library (or whatever wrapper we write) won't have the capacity to differentiate between the infinite loop quasi-state and a temporarily blocked state.

I think it makes more sense to comply with the backoff library's semantics.

I'm happy to followup with a separate PR to fix the other issue.

**Comment by advait** [comment]
Repharased: the backoff library (or any backoff mechanism) can't differentiate between:
1. A goroutine that's blocked and has neither succeeded nor failed yet
2. A goroutine that is "successfully" doing work in an infinite loop

We need some mechanism to indicate success to the backoff mechanism for it to reset its backoff time, etc.

**Comment by emilevauge** [comment]
@advait Yeah, you are probably right.
I was hoping to build a specific exponential backoff function to be used with long running jobs. Maybe we could just disable `MaxElapsedTime` setting it to `0`. But it would not detect blocked jobs.
I don't know what you think but in that case of long-running jobs, I think it's reasonable to disable this parameter.

**Comment by emilevauge** [comment]
I think we can close this one in favor of #627 ;)

**Pull request closed without merging**